internal/venice-common/src/main/java/com/linkedin/venice/writer/ComplexVeniceWriter.java

```diff
@@ -0,0 +1,163 @@
+        if (partitions.length == 0) {
+          finalCompletableFuture.complete(null);
+        } else {
+          byte[] serializedValue = valueSerializer.serialize(topicName, value);
+          performMultiPartitionAction(
+              partitions,
+              finalCompletableFuture,
+              (partition) -> this.put(serializedKey, serializedValue, valueSchemaId, partition));
+        }
+      }
+    }
+    return finalCompletableFuture;
+  }
+
+  /**
+   * Perform "delete" on the given key. If a {@link ComplexVenicePartitioner} is involved then it will be a best effort
+   * attempt to delete the record using the valueProvider. It's best effort because:
+   *   1. Nothing we can do if value is null or not provided via valueProvider.
+   *   2. Previous writes
+   */
+  public CompletableFuture<Void> complexDelete(K key, Lazy<GenericRecord> valueProvider) {
+    CompletableFuture<Void> finalCompletableFuture = new CompletableFuture<>();
+    if (complexPartitioner == null) {
+      // No VeniceComplexPartitioner involved, perform simple delete.
+      propagateVeniceWriterFuture(this.delete(key, null), finalCompletableFuture);
+    } else {
+      GenericRecord value = valueProvider.get();
+      if (value == null) {
+        // Ignore the delete since we cannot perform delete with VeniceComplexPartitioner without the value
+        finalCompletableFuture.complete(null);
+      } else {
+        byte[] serializedKey = keySerializer.serialize(topicName, key);
+        int[] partitions = complexPartitioner.getPartitionId(serializedKey, value, numberOfPartitions);
+        if (partitions.length == 0) {
+          finalCompletableFuture.complete(null);
+        } else {
+          performMultiPartitionAction(
+              partitions,
+              finalCompletableFuture,
+              (partition) -> this.delete(serializedKey, null, partition));
+        }
+      }
+    }
+    return finalCompletableFuture;
+  }
+
+  /**
+   * Prevent the {@link ComplexVeniceWriter} from writing any actual chunks for large values. This is because we are
+   * only using ComplexVeniceWriter for writing to materialized view. The consumers of materialized view do not fully
+   * support assembling the chunks correctly yet. The behavior is the same for both large values from VPJ and leader
+   * replicas.
+   */
+  @Override
+  protected CompletableFuture<PubSubProduceResult> putLargeValue(
+      byte[] serializedKey,
+      byte[] serializedValue,
+      int valueSchemaId,
+      PubSubProducerCallback callback,
+      int partition,
+      LeaderMetadataWrapper leaderMetadataWrapper,
+      long logicalTs,
+      PutMetadata putMetadata,
+      ChunkedValueManifest oldValueManifest,
+      ChunkedValueManifest oldRmdManifest) {
+    skippedLargeRecords.incrementAndGet();
+    if (!REDUNDANT_LOGGING_FILTER.isRedundantException(topicName, SKIP_LARGE_RECORD)) {
```

Comment:
Do we need to check whether it is a complex partitioner or not before skipping?
For the regular partitioner, I assume we can let large value go to viewer topic, right?

Reply:
No, we still cannot let large value to go to viewer topic even for regular partitioner. This is because even for regular partitioner a single partition can have multiple leader writing to it and breaks the assumption of single leader for chunk assembling. The consumer support for that is still missing.

Reply:
This is bad...
We do have DaVinci use cases, who might want to use the regular partitioner and DaVinci can handle chunking in a correct way.
I wondered whether the large message skipping can be done in CDC instead of here for regular partitioner.
It can be done in a later PR, but I do think we need to revaluate such decision.

Reply:
Sure, we can revisit this but wanted to point out that even for DVC chunking is not fully supported today if we have large values in the batch push. This is because during NR pass-through we don't know how to forward the chunks to view topic partitions correctly. The logic as of now will partition and send the chunks as regular records which will most likely be wrong.